restapi/src/main/java/io/stargate/web/restapi/dao/RestDBAccessFactory.java

```diff
@@ -0,0 +1,123 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.stargate.web.restapi.dao;
+
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.LoadingCache;
+import io.stargate.auth.AuthenticationService;
+import io.stargate.auth.AuthenticationSubject;
+import io.stargate.auth.AuthorizationService;
+import io.stargate.auth.UnauthorizedException;
+import io.stargate.db.datastore.DataStore;
+import io.stargate.db.datastore.DataStoreFactory;
+import io.stargate.db.datastore.DataStoreOptions;
+import java.time.Duration;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.CompletionException;
+import java.util.stream.Collectors;
+
+public class RestDBAccessFactory {
+  private final AuthenticationService authenticationService;
+  private final AuthorizationService authorizationService;
+
+  private final LoadingCache<TokenAndHeaders, RestDBAccess> restTokensToDB =
+      Caffeine.newBuilder()
+          .maximumSize(10_000)
+          .expireAfterWrite(Duration.ofMinutes(1))
+          .build(this::getRestDataStoreForTokenInternal);
+
+  private final DataStoreFactory dataStoreFactory;
+
+  public RestDBAccessFactory(
+      AuthenticationService authenticationService,
+      AuthorizationService authorizationService,
+      DataStoreFactory dataStoreFactory) {
+    this.authenticationService = authenticationService;
+    this.authorizationService = authorizationService;
+    this.dataStoreFactory = dataStoreFactory;
+  }
+
+  public RestDBAccess getRestDBForToken(String token, Map<String, String> headers)
+      throws UnauthorizedException {
+    if (token == null) {
+      throw new UnauthorizedException("Missing token");
+    }
+
+    try {
+      return restTokensToDB.get(TokenAndHeaders.create(token, headers));
+    } catch (CompletionException e) {
+      if (e.getCause() instanceof UnauthorizedException) {
+        throw (UnauthorizedException) e.getCause();
+      }
+      throw e;
+    }
+  }
+
+  private RestDBAccess getRestDataStoreForTokenInternal(TokenAndHeaders tokenAndHeaders)
+      throws UnauthorizedException {
+    AuthenticationSubject authenticationSubject =
+        authenticationService.validateToken(tokenAndHeaders.token, tokenAndHeaders.headers);
+    return new RestDBAccess(
+        constructDataStore(authenticationSubject, tokenAndHeaders),
+        authenticationSubject,
+        authorizationService);
+  }
+
+  private DataStore constructDataStore(
+      AuthenticationSubject authenticationSubject, TokenAndHeaders tokenAndHeaders) {
+    return dataStoreFactory.create(
+        authenticationSubject.asUser(),
+        DataStoreOptions.builder()
+            .alwaysPrepareQueries(true)
+            .putAllCustomProperties(tokenAndHeaders.headers)
+            .build());
+  }
+
+  static class TokenAndHeaders {
```

Comment:
we have this duplicated now? extract, abstract class?

Reply:
If we are to split Document and REST APIs, they should not share anything just amongst each other: either there is bit of duplication, or we extract common implementation into something that may be shared by other services, I think.
This particular inner class is not exposed outside at all so we could go either way.

My thinking here would be to first use bit of duplication, but then extract reusable class into some common component. I just don't know where -- worth a discussion?